### ISSUE TITLE
Unable to reproduce the results on the Camelyon16 dataset.

### ISSUE THREAD

**Issue by Henry0394** [DearCaat/MHIM-MIL#20]
I'm attempting to reproduce the results on the Camelyon16 dataset, but I'm obtaining an accuracy rate of around 78%-80%, which is approximately 10 percent lower than the reported accuracy in the paper. Could there be some critical details I've overlooked? Or could there potentially be a bug in the code?

**Comment by DearCaat** [comment]
It would be helpful to first reproduce the results of classic `ABMIL` to check your environment and training scripts.

**Comment by Henry0394** [comment]
> It would be helpful to first reproduce the results of classic `ABMIL` to check your environment and training scripts.

Thanks, and I did, but the result for ABMIL is still under 80%, it is really weird.

**Comment by DearCaat** [comment]
That may require double-check the dataset, the experimental environment, and the training codes.

**Comment by Henry0394** [comment]
Thanks, but I'm not the only one unable to reproduce the results. As far as I know, several others I'm familiar with have also encountered similar issues. Are there any tricks that aren't mentioned in the code, or could there be bugs within the codebase?

**Comment by DearCaat** [comment]
Although I would like to share some secret tips with you, the Docker and code I provided are exactly what I used. The code in this repository has been refactored, so it may not reproduce the results of the paper with 100% accuracy, and there might be a deviation of ±1-2%. However, if the deviation exceeds 10%, or if AB-MIL fails to produce valid results, I would find it very strange. If you are using the data, Docker, and code I provided and still see a deviation of over 10%, could you please share the training logs? I will do my best to help resolve the issue.

**Comment by BUPT-BownZ** [comment]
Hello, I've encountered the same issue. Three months ago, I reproduced this article, and at that time, the evaluation metrics on ABmil were roughly the same as those in the paper. However, when I run the program again now, the metrics for ABmil-MHIM on the C16 dataset are also unable to reach an 80% accuracy rate. During this period, I haven't changed any code. Could you please help me understand what might have caused this issue?

**Comment by BUPT-BownZ** [comment]
This issue is also the same when running ABmil-MHIM on the TCGA dataset.

**Comment by DearCaat** [comment]
> Hello, I've encountered the same issue. Three months ago, I reproduced this article, and at that time, the evaluation metrics on ABmil were roughly the same as those in the paper. However, when I run the program again now, the metrics for ABmil-MHIM on the C16 dataset are also unable to reach an 80% accuracy rate. During this period, I haven't changed any code. Could you please help me understand what might have caused this issue?

If you are using the data, Docker, and code I provided and still see a deviation of over 10%, could you please share the training command, logs and initialization weights for the teacher model? I will do my best to help resolve the issue.

**Comment by BUPT-BownZ** [comment]
是的，在TCGA数据集上的表现也很差，这个图上的结果是今天我运行代码后得到的结果。
![image](https://github.com/user-attachments/assets/36d1dcc4-4f46-46a5-9216-85f6785a007a)
下面这张是我三个月前在TCGA数据上运行ABmil-MHIM得到的结果
![image](https://github.com/user-attachments/assets/6785365e-c52f-42da-96a0-c5582c2f895d)
在teacher模型上我选取的还是之前三个月前训练好得到的的baseline模型——**fold_2_model_best_auc.pt**
关于ABmil-MHIM具体我的执行命令是
python3 main.py \
    --project=tcga_lung_mhim2\
    --dataset_root=./datasets/tcga \
    --model_path=./output \
    --cv_fold=4 \
    --val_ratio=0.13 \
    --teacher_init=./output/tcga_lung_baseline/abmil/fold_2_model_best_auc.pt \
    --title="abmil_105_mr70l20h2-0_mmcos_is" \
    --baseline=attn \
    --num_workers=0 \
    --cl_alpha=0.5 \
    --mask_ratio_h=0.02 \
    --mrh_sche \
    --mm_sche \
    --init_stu_type=fc \
    --mask_ratio=0.7 \
    --mask_ratio_l=0.2 \
    --seed=2021 \
    --num_workers=0 \
    --datasets=tcga

**Comment by DearCaat** [comment]
> 是的，在TCGA数据集上的表现也很差，这个图上的结果是今天我运行代码后得到的结果。 ![image](https://private-user-images.githubusercontent.com/144535995/381215352-36d1dcc4-4f46-46a5-9216-85f6785a007a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzAyNTU5MzAsIm5iZiI6MTczMDI1NTYzMCwicGF0aCI6Ii8xNDQ1MzU5OTUvMzgxMjE1MzUyLTM2ZDFkY2M0LTRmNDYtNDZhNS05MjE2LTg1ZjY3ODVhMDA3YS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDMwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAzMFQwMjMzNTBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hMmQ4YWIwM2E2OWQ5M2E5ODQ2ZGM0NzgxZTBjMGQzMDQ2NjVjODZkMDllMDEwYTMyNDg3YTgzMWZiYjEzZGIxJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.ntiUDcs0_eBy7ZnMdPHw20C9BQgwLJpkfpLGRdZZYzU) 下面这张是我三个月前在TCGA数据上运行ABmil-MHIM得到的结果 ![image](https://private-user-images.githubusercontent.com/144535995/381215912-6785365e-c52f-42da-96a0-c5582c2f895d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzAyNTU5MzAsIm5iZiI6MTczMDI1NTYzMCwicGF0aCI6Ii8xNDQ1MzU5OTUvMzgxMjE1OTEyLTY3ODUzNjVlLWM1MmYtNDJkYS05NmEwLWM1NTgyYzJmODk1ZC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDMwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAzMFQwMjMzNTBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT03OTdjMThhMmNmMTFhZjc2ZTM5ZjUxODU3ZGE1ZWQ5YzkzMmNkZmI3MjczMzA2ZDQxYmVkMjU2YTNlNjY3NzkyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.aQ8GBeA0NLkw1vfhPUno8dRe3b-xA5aajTuDh8uOEhA) 在teacher模型上我选取的还是之前三个月前训练好得到的的baseline模型——**fold_2_model_best_auc.pt** 关于ABmil-MHIM具体我的执行命令是 python3 main.py --project=tcga_lung_mhim2 --dataset_root=./datasets/tcga --model_path=./output --cv_fold=4 --val_ratio=0.13 --teacher_init=./output/tcga_lung_baseline/abmil/fold_2_model_best_auc.pt --title="abmil_105_mr70l20h2-0_mmcos_is" --baseline=attn --num_workers=0 --cl_alpha=0.5 --mask_ratio_h=0.02 --mrh_sche --mm_sche --init_stu_type=fc --mask_ratio=0.7 --mask_ratio_l=0.2 --seed=2021 --num_workers=0 --datasets=tcga

- 从logit_loss来看，模型应该没有任何的训练，不知道能不能检查关于CE loss的优化情况
- 建议先不使用任何`teacher_init`的情况下再进行测试，而且因为是多折交叉验证，不能直接使用其中某一折（`fold2`）来初始化所有折的模型，这样会导致测试集泄露的问题
- 从log中的warning可以看出，似乎prediction也是有问题的，不知道能不能检查一下
- 还有可能就是数据集本身的label是否加载成功，不知道能否检查[代码](https://github.com/DearCaat/MHIM-MIL/blob/master/dataloader.py#L138)，在这之后，打印出训练的label分布。鉴于你说的你完全没有更改代码，我怀疑可能是数据集label的问题，或者数据集特征读取的问题

**Comment by BUPT-BownZ** [comment]
确实是loss函数中存在一些问题，现在重新配置环境已经解决了，感谢作者！

**Comment by DearCaat** [comment]
> 确实是loss函数中存在一些问题，现在重新配置环境已经解决了，感谢作者！

好的，感谢对我们工作的关注，如果有任何问题欢迎继续在该issue上提问，我尽力回答

**Comment by weiaicunzai** [comment]
> 确实是loss函数中存在一些问题，现在重新配置环境已经解决了，感谢作者！

loss函数中有啥问题吗？能分享一下吗？我也遇到acc上不去的情况。

**Comment by weiaicunzai** [comment]
> > 确实是loss函数中存在一些问题，现在重新配置环境已经解决了，感谢作者！
> 
> 好的，感谢对我们工作的关注，如果有任何问题欢迎继续在该issue上提问，我尽力回答

请问你有试过把github的这份代码下载下来，然后重新配置环境训练吗？

**Comment by DearCaat** [comment]
> > > 确实是loss函数中存在一些问题，现在重新配置环境已经解决了，感谢作者！
> > 
> > 
> > 好的，感谢对我们工作的关注，如果有任何问题欢迎继续在该issue上提问，我尽力回答
> 
> 请问你有试过把github的这份代码下载下来，然后重新配置环境训练吗？

不好意思，请问你还有这个问题嘛，如果有的话，我近期会再次调试github的代码

**Comment by DearCaat** [comment]
> > > 确实是loss函数中存在一些问题，现在重新配置环境已经解决了，感谢作者！
> > 
> > 
> > 好的，感谢对我们工作的关注，如果有任何问题欢迎继续在该issue上提问，我尽力回答
> 
> 请问你有试过把github的这份代码下载下来，然后重新配置环境训练吗？

如果是ABMIL训练的结果都不理想，推荐先使用https://github.com/DearCaat/RRT-MIL/tree/master 以及https://github.com/lingxitong/MIL_BASELINE 这两个代码库进行尝试。当前github上的版本比较陈旧，需要进一步更新不好意思
